skcosmo/preprocessing/KernelFlexibleCenterer.py

```diff
@@ -0,0 +1,85 @@
+import numpy as np
+import sklearn as sk
+from sklearn.base import TransformerMixin, BaseEstimator
+
+
+class KernelFlexibleCenterer(TransformerMixin, BaseEstimator):
+    """Kernel centering method, similar to KernelCenterer,
+    but with additional parameters, relative to which centering
+    is carried out:
+    ref_cmean: an array with means for each column.
+    ref_rmean: an array with means for each row.
+    ref_mean: an average for the whole kernel matrix
+    reference: a reference, relative to which centering is carried out"""
+
+    def __init__(
+        self, *, reference=None, ref_cmean=None, ref_rmean=None, ref_mean=None
+    ):
+        """Initialize KernelFlexibleCenterer. Specifies the parameters relative to which the  kernel
+        matrix will be centered."""
+
+        self.reference = reference
+        self.ref_cmean = ref_cmean
+        self.ref_rmean = ref_rmean
+        self.ref_mean = ref_mean
+
+    def fit(self, K, y=None):
+        """Fit KernelFlexibleCenterer
+        Parameters
+        ----------
+        K : ndarray of shape (n_samples, n_samples)
+            Kernel matrix.
+        y : None
+            Ignored.
+        Returns
+        -------
+        self : object
+            Fitted transformer.
+        """
+        K_ref = self.reference
+        if K_ref is None:
+            K_ref = K
+            self.reference = K
+        else:
+            if K.shape[1] != K_ref.shape[0]:
+                raise ValueError(
+                    "The kernel to be centered and the reference have inconsistent sizes"
+                )
+        if K_ref.shape[0] != K_ref.shape[1]:
+            raise ValueError(
+                "The reference kernel is not square, and does not define a RKHS"
+            )
+
+        if self.ref_cmean is None:
+            self.ref_cmean = K_ref.mean(axis=0)
+
+        if self.ref_rmean is None:
+            self.ref_rmean = K_ref.mean(axis=1)
+
+        if self.ref_mean is None:
+            self.ref_mean = K_ref.mean()
+        return self
+
+    def transform(self, K, y=None):
+        """ Centering our Kernel. Previously you should fit data. """
+        params = self.get_params()
+        for key in params.keys():
+            if params[key] is None:
+                raise sk.exceptions.NotFittedError(
+                    "This "
+                    + type(self).__name__
+                    + " instance is not fitted yet. Call 'fit' with appropriate arguments before using this estimator."
+                )
+        Kc = (
+            K
+            - np.broadcast_arrays(K, self.ref_cmean)[1]
+            - self.ref_rmean.reshape((K.shape[0], 1))
+            + np.broadcast_arrays(K, self.ref_mean)[1]
+        )
+
+        return Kc
+
+    def fit_transform(self, K, y=None, **fit_params):
```

Comment:
needs doc

Reply:
where do the `fit_params` come into play?

Reply:
This is necessary for compatibility with the functions of the TransformerMixin class

Reply:
hm, is it necessary?

Reply:
Yes, I tried to write without it, receive an error